File: backend/src/Loupe.Application/PhotographerDrafts/PhotographerDraftResult.cs
using Loupe.Application.Operations;
using Loupe.Domain.Photographers;
using System.Text.Json;

namespace Loupe.Application.PhotographerDrafts;

public sealed record PhotographerDraftResult(Guid Id, string PortfolioUrl, string? Name, long Revision, DateTimeOffset ExpiresAt,
    Guid? CommittedPhotographerId, string? FailureCode, OperationResult? Import, string? Description, string[] Tags, CapturedPortfolioPage? Source)
{
    public static PhotographerDraftResult From(PhotographerDraft draft)
    {
        var source = draft.SourceJson is null ? null : JsonSerializer.Deserialize<CapturedPortfolioPage>(draft.SourceJson);
        return new(draft.Id, draft.PortfolioUrl, draft.Name, draft.Revision,
        draft.ExpiresAt, draft.CommittedPhotographerId, draft.FailureCode ?? draft.ImportOperation?.FailureCode,
        draft.ImportOperation is null ? null : OperationResult.From(draft.ImportOperation), source?.Description, source?.Tags ?? [], source);
    }
}

File: frontend/projects/api/src/lib/photographer-draft/photographer-draft-result.ts
import { OperationResult } from '../operation/operation-result';
import { PhotographerMetadata } from '../photographer/photographer-metadata';

export interface PhotographerDraftResult {
  id: string;
  portfolioUrl: string;
  name: string | null;
  description: string | null;
  tags: string[];
  revision: number;
  expiresAt: string;
  committedPhotographerId: string | null;
  failureCode: string | null;
  import: OperationResult | null;
}
export interface SavePhotographerDraftInput extends PhotographerMetadata {
  revision: number;
}
